### ISSUE TITLE
[arrow-parens] Add a "when-brace" option

### ISSUE THREAD

**Issue by nfroidure** [eslint/eslint#6557]
Our team would like to adopt the AirBnb best practice for arrow functions parens but it appears the current options of this rule doesn't fit it. Since we want to rely on automated coding style rules only and found out it would be useful for others too, here is the PR that fulfill it.

See https://github.com/airbnb/javascript#arrows--one-arg-parens
Discussion: https://github.com/SimpliField/eslint-config-simplifield/issues/14

- The version of ESLint you are using

2.11.1
https://github.com/SimpliField/eslint-config-simplifield/blob/master/package.json#L20

- The rule you want to change

'arrow-parens'

- The code you want to be flagged as incorrect
```js
a => { a } // (1)
(a) => a // (2)
```
- What happens when the rule is applied to the code without your change

with 'as-needed' (1) is valid

with 'always' (2) is valid

**Comment by platinumazure** [comment]
In order to facilitate acceptance of this issue and pull request, could you please edit your original post to answer the questions raised in the [Proposing a Rule Change](http://eslint.org/docs/developer-guide/contributing/rule-changes) documentation? Thanks!

**Comment by nfroidure** [comment]
@platinumazure done, missed that part.

**Comment by ilyavolodin** [comment]
Since this is part of AirBnB styleguide, I think we should consider adding an option for this. I don't like proposed name, however. Don't really have any ideas on how to name it though. I'll :+1: this

**Comment by nfroidure** [comment]
Agree on the naming, not convinced by it too but nothing better came to me ;).

**Comment by ilyavolodin** [comment]
@eslint/eslint-team We need some more +1s here.

**Comment by kaicataldo** [comment]
:+1:

**Comment by mikesherov** [comment]
👍

**Comment by IanVS** [comment]
How about breaking this up a bit, and adding an option object with keys like `one-arg` and `with-braces`?  The values for these could be `always`, or `as-needed`.  I don't love this and I'd prefer to make them booleans, but I'm not sure how to do that in a way that makes sense.  Hopefully this comment might spark a better idea.

The upshot would be making this a bit more flexible and still accommodating the AirBnB style mentioned by the OP. (`{'one-arg': 'as-needed', 'with-braces': 'always'}`)

**Comment by nzakas** [comment]
Do we have a champion for this?

I'd suggest we look at a configuration like:

```
arrow-parens: [ "error", "as-needed", {
    requireForBlockBody: true
}]
```

(Generally I think "block" is a more appropriate description, but could also be `requireForBracedBody`.)

**Comment by IanVS** [comment]
@nzakas, I think we would also need `requireForMultipleArguments`, to match AirBnB style, but I like that strategy.

**Comment by nzakas** [comment]
@ianvs hmmm, I'm not sure I understand. You can't have multiple parameters for an arrow function without parentheses (it's a syntax error). Can you explain a bit more?

**Comment by IanVS** [comment]
Doh, of course.  I was just reading the styleguide way too literally and trying to invert it without thinking about it:

> If your function takes a single argument and doesn’t use braces, omit the parentheses.

**Comment by kaicataldo** [comment]
I can champion this!

**Comment by alberto** [comment]
👍  to `requireForBlockBody`

**Comment by nzakas** [comment]
Okay @kaicataldo, the power is yours!

**Comment by kaicataldo** [comment]
Thanks! Have we decided that `requireForBlockBody` is config option name we want?

**Comment by nzakas** [comment]
Yup!
